src/function/cast/variant/from_variant.cpp

```diff
@@ -0,0 +1,659 @@
+#include "duckdb/function/cast/default_casts.hpp"
+#include "duckdb/common/types/variant.hpp"
+#include "duckdb/function/scalar/variant_utils.hpp"
+#include "duckdb/common/serializer/varint.hpp"
+#include "yyjson.hpp"
+
+#include "duckdb/common/enum_util.hpp"
+#include "duckdb/common/exception/conversion_exception.hpp"
+
+using namespace duckdb_yyjson; // NOLINT
+
+namespace duckdb {
+
+namespace {
+
+struct FromVariantConversionData {
+public:
+	explicit FromVariantConversionData(RecursiveUnifiedVectorFormat &variant_format) : variant(variant_format) {
+	}
+
+public:
+	//! The input Variant column
+	UnifiedVariantVectorData variant;
+	//! If unsuccessful - the error of the conversion
+	string error;
+};
+
+struct EmptyConversionPayloadFromVariant {};
+
+//! string data
+struct StringConversionPayload {
+public:
+	explicit StringConversionPayload(Vector &vec) : vec(vec) {
+	}
+
+public:
+	//! The string vector that needs to own the non-inlined data
+	Vector &vec;
+};
+
+//! decimal
+struct DecimalConversionPayloadFromVariant {
+public:
+	DecimalConversionPayloadFromVariant(idx_t width, idx_t scale) : width(width), scale(scale) {
+	}
+
+public:
+	idx_t width;
+	idx_t scale;
+};
+
+struct ConvertedJSONHolder {
+public:
+	~ConvertedJSONHolder() {
+		if (doc) {
+			yyjson_mut_doc_free(doc);
+		}
+		if (stringified_json) {
+			free(stringified_json);
+		}
+	}
+
+public:
+	yyjson_mut_doc *doc = nullptr;
+	char *stringified_json = nullptr;
+};
+
+} // namespace
+
+//===--------------------------------------------------------------------===//
+// VARIANT -> ANY
+//===--------------------------------------------------------------------===//
+
+static bool FinalizeErrorMessage(FromVariantConversionData &conversion_data, Vector &result,
+                                 CastParameters &parameters) {
+	auto conversion_error = StringUtil::Format("%s to '%s'", conversion_data.error, result.GetType().ToString());
+	if (parameters.error_message) {
+		*parameters.error_message = conversion_error;
+		return false;
+	}
+	throw ConversionException(conversion_error);
+}
+
+//! ------- Primitive Conversion Methods -------
+
+//! bool
+struct VariantBooleanConversion {
+	using type = bool;
+	static bool Convert(const VariantLogicalType type_id, uint32_t byte_offset, const_data_ptr_t value, bool &ret,
+	                    const EmptyConversionPayloadFromVariant &payload, string &error) {
+		if (type_id != VariantLogicalType::BOOL_FALSE && type_id != VariantLogicalType::BOOL_TRUE) {
+			error = StringUtil::Format("Can't convert from VARIANT(%s)", EnumUtil::ToString(type_id));
+			return false;
+		}
+		ret = type_id == VariantLogicalType::BOOL_TRUE;
+		return true;
+	}
+};
+
+//! any direct conversion (int8, date_t, dtime_t, timestamp, etc..)
+template <class T, VariantLogicalType TYPE_ID>
+struct VariantDirectConversion {
```

Comment:
This only supports direct conversion (INT8 -> INT8 for example), when the types differ we fall back to a slow `Value.DefaultCastAs(target_type)` conversion, to ensure feature completeness, at the cost of performance.
Over time this can (should) be optimized to extend the casting logic to support differing types, at the risk of reimplementing all the existing cast logic.

^ Future work